### PR TITLE
WIP: minimalistic Monitor 

### PR DESCRIPTION
**Pull request by Liamsi**
~~This will be split up into small separate PRs~~ (related to #679).
~~As discussed with Gary this will be merge as a single PR.~~

After further offline discussion this will be split up into two parts:

1) A monitor "skeleton" which does everything besides the verification
2) the verification logic and other changes to core (like to the protos)

-------

Implemented:
- fetch mutations from server
- main function
- Docker/docker-compose
- basic HTTP/grpc API (will be renamed according to [this](https://cloud.google.com/apis/design/resources))
- signing and publishing result/"gossiping"
- verification (of mutations, SMHs); untested 

Missing:
- "cleanup" (`impl` vs `core`)
- proper storage of received SMH

### CODE REVIEW

File: cmd/keytransparency-monitor/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ENV MAP_ID=0 \
+    MAP_URL=""
+ENV LOG_ID=0 \
+    LOG_URL=localhost:8090
+ENV MONITOR_SIGN_KEY=TODO \
+    POLL_PERIOD=5s
+
+ENV VERBOSITY=1
+
+ADD keytransparency/genfiles/* /kt/
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/Dockerfile]: Try following rob's example

File: cmd/keytransparency-monitor/main.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	vrfPath   = flag.String("vrf", "genfiles/vrf-pubkey.pem", "Path to VRF public key")
+	logPubKey = flag.String("log-key", "genfiles/trillian-log.pem", "Path to the trillian log's public key")
+	keyFile   = flag.String("key", "genfiles/server.key", "TLS private key file")
+	certFile  = flag.String("cert", "genfiles/server.pem", "TLS cert file")
+
+	pollPeriod = flag.Duration("poll-period", time.Second*5, "Maximum time between polling the key-server. Ideally, this is equal to the min-period of paramerter of the keyserver.")
+	ktURL      = flag.String("kt-url", "localhost:8080", "URL of key-server.")
+	ktPEM      = flag.String("kt-key", "genfiles/server.crt", "Path to kt-server's public key")
+	// TODO(ismail): are the IDs actually needed for verification operations?
+	mapID = flag.Int64("map-id", 0, "Trillian map ID")
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: you can remove mapID and logID

File: cmd/keytransparency-monitor/main.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	"github.com/golang/glog"
+	"github.com/google/keytransparency/impl/monitor"
+	"github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/reflection"
+)
+
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: Add TODO for a prometheus variable that tracks Valid / Invalid for each Tree Head

File: core/proto/keytransparency_v1_types/gen.go [hunk 4]
```diff
@@ -12,6 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//go:generate protoc -I=. -I=$GOPATH/src/ -I=$GOPATH/src/github.com/google/trillian/ -I=$GOPATH/src/github.com/googleapis/googleapis --go_out=:. keytransparency_v1_types.proto
+//go:generate protoc -I=. -I=$GOPATH/src/ -I=$GOPATH/src/github.com/google/trillian/vendor/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis/ -I=$GOPATH/src/github.com/google/trillian/ --go_out=:. keytransparency_v1_types.proto
```
Comment (gdbelvin) [inline core/proto/keytransparency_v1_types/gen.go]: revert local changes

Comment (gdbelvin) [inline core/proto/keytransparency_v1_types/gen.go]: try rebasing

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto [hunk 5]
```diff
@@ -236,3 +236,19 @@ message GetMutationsResponse {
...
+
+// GetMonitoringRequest contains the input parameters of the GetMonitoring APIs.
+message GetMonitoringRequest {
+  // start specifies the start epoch number from which monitoring results will
+  // be returned (ranging from [start, latestObserved] and starting at 1).
+  int64 start = 1;
+}
+
+message GetMonitoringResponse {
+  // smr contains the map root for the sparse Merkle Tree signed with the
```
Comment (gdbelvin) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: For convenience: a pass / fail field. 

Proof fields for each type of failure.

File: impl/monitor/monitor.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+// GetSignedMapRoot returns the latest reconstructed using the Mutations API and
+// validated signed map root.
+func (s *server) GetSignedMapRoot(ctx context.Context, in *ktpb.GetMonitoringRequest) (*ktpb.GetMonitoringResponse, error) {
+	return &ktpb.GetMonitoringResponse{
+		Smr: s.lastSMR(),
+	}, nil
+}
+
+// GetSignedMapRootStream is a streaming API similar to GetSignedMapRoot.
+func (s *server) GetSignedMapRootStream(in *ktpb.GetMonitoringRequest, stream mspb.MonitorService_GetSignedMapRootStreamServer) error {
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Remove functions that you don't want to implement. -- including from proto

File: impl/monitor/monitor.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+		return nil, err
+	}
+
+	// TODO(ismail): remember when we've actually requested and seen this SMR!
+	// publish delta (seen - actual)
+	if got, want := resp.GetSmr(), s.lastSeenSMR(); bytes.Equal(got.GetRootHash(), want.GetRootHash()) &&
+		got.GetMapRevision() == want.GetMapRevision() {
+		// We already processed this SMR. Do not update seen SMRs. Do not scroll
+		// pages for further mutations. Return empty mutations list.
+		glog.Info("Already processed this SMR. Continuing.")
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: include revision number

File: impl/monitor/monitor.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		return nil, nil
+	}
+
+	mutations := make([]*ktpb.Mutation, pageSize*2)
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Put the code to iterate through pages in a function

File: impl/monitor/monitor.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+	// update seen SMRs:
+	s.seenSMRs = append(s.seenSMRs, resp.GetSmr())
+
+	// TODO(ismail):
+	// For each received mutation in epoch e:
+	// Verify that the provided leaf’s inclusion proof goes to epoch e -1.
+	// Verify the mutation’s validity against the previous leaf.
+	// Compute the new leaf and store the intermediate hashes locally.
+	// Compute the new root using local intermediate hashes from epoch e.
+	for _, m := range mutations {
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Put verification code in it's own function too

File: impl/monitor/monitor.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+const pageSize = 16
+
+// Server holds internal state for the monitor server.
+type server struct {
+	client     mupb.MutationServiceClient
+	pollPeriod time.Duration
+
+	tree *tv.Verifier
+	// TODO(ismail) abstract this into a storage interface and have an in-memory version
+	seenSMRs          []*trillian.SignedMapRoot
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Maybe create a struct for this:
- SMR
- Was it valid? / Error type / status
- Time observed

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto (hunk 5 above)
Comment (Liamsi) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: Thanks! I've updated your comment to clarify that the "invalid mutations" case is only from epoch e to e+1. My understanding still is that the "storage-less" monitor doesn't have all information if it doesn't additionally store the inclusion proofs at each epoch e. As the monitor can not know in advance (at epoch e) if there will be a problem in e+1. Is that correct or do I miss sth?

File: cmd/keytransparency-monitor/main.go (hunk 2 above)
Comment (Liamsi) [inline cmd/keytransparency-monitor/main.go]: `mapID` is currently needed by the [verifier](https://github.com/google/keytransparency/blob/142a834769e003a154841f760511550e4c25fd2c/core/tree/sparse/verifier/verifier.go#L41). I've added a TODO to remove it as soon as the verifier is updated (similar to https://github.com/google/keytransparency/pull/708)

File: cmd/keytransparency-monitor/Dockerfile (hunk 1 above)
Comment (Liamsi) [inline cmd/keytransparency-monitor/Dockerfile]: Done. I opened an issue for our other Dockerfiles (#712)

File: core/proto/keytransparency_v1_types/gen.go (hunk 4 above)
Comment (Liamsi) [inline core/proto/keytransparency_v1_types/gen.go]: done

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto (hunk 5 above)
Comment (Liamsi) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: @gdbelvin I've added documenting this in a markdown document to the issue's TODO list so we won't forget about it.

Note to myself: the new inclusion proof (e+1) can be computed using the mutations and the old inclusion proof.

Comment (Liamsi) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: I've added proto messages for the above cases here: https://github.com/google/keytransparency/pull/709/commits/5ed5f8a63dcdef8b374c79ea6ce4cc29e0304576

Let me know what you think.

File: impl/monitor/monitor.go (hunk 7 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: done

File: impl/monitor/monitor.go (hunk 8 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: done

File: impl/monitor/monitor.go (hunk 9 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: done. though, the verification code does not really verify anything yet

File: cmd/keytransparency-monitor/main.go [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+package main
+
+import (
+	"flag"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	mopb "github.com/google/keytransparency/impl/proto/monitor_v1_service"
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: please place this import group at the bottom of `import(`

File: cmd/keytransparency-monitor/main.go [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/reflection"
+)
+
+var (
+	addr     = flag.String("addr", ":8099", "The ip:port combination to listen on")
+	keyFile  = flag.String("key", "genfiles/server.key", "TLS private key file")
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: tls-key and tls-cert?

File: cmd/keytransparency-monitor/main.go [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+	certFile = flag.String("cert", "genfiles/server.pem", "TLS cert file")
+
+	signingKey         = flag.String("sign-key", "genfiles/p256-key.pem", "Path to private key PEM for SMH signing")
+	signingKeyPassword = flag.String("password", "towel", "Password of the private key PEM file for SMH signing")
+
+	pollPeriod = flag.Duration("poll-period", time.Second*5, "Maximum time between polling the key-server. Ideally, this is equal to the min-period of paramerter of the keyserver.")
+	ktURL      = flag.String("kt-url", "localhost:8080", "URL of key-server.")
+	ktPEM      = flag.String("kt-key", "genfiles/server.crt", "Path to kt-server's public key")
+	// TODO(ismail): remove mapID
+	mapID = flag.Int64("map-id", 0, "Trillian map ID")
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: You will need a mapKey

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto [hunk 14]
```diff
@@ -236,3 +236,63 @@ message GetMutationsResponse {
...
+  repeated trillian.MapLeafInclusion new_leafs = 3;
+}
+
+message GetMonitoringResponse {
+  // smr contains the map root for the sparse Merkle Tree signed with the
+  // monitor's key on success. If the checks were not successful the
+  // smr will be empty. The epochs are encoded into the smr map_revision.
+  trillian.SignedMapRoot smr = 1;
+
+  // seen_timestamp_nanos contains the time in nanoseconds where this particular
```
Comment (gdbelvin) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: We might need to create a signed version of this, but let's discuss that in the design document

File: impl/monitor/monitor.go [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+// If the monitor could not reconstruct the map root given the set of
+// mutations from the previous to the current epoch it won't sign the map root
+// and additional data will be provided to reproduce the failure.
+func (s *Server) GetSignedMapRootByRevision(ctx context.Context, in *ktpb.GetMonitoringRequest) (*ktpb.GetMonitoringResponse, error) {
+	// TODO(ismail): implement by revision API
+	return nil, grpc.Errorf(codes.Unimplemented, "GetSignedMapRoot is unimplemented")
+}
+
+func (s *Server) pollMutations(ctx context.Context, opts ...grpc.CallOption) ([]*ktpb.Mutation, error) {
+	req := &ktpb.GetMutationsRequest{PageSize: pageSize, Epoch: s.nextRevToQuery()}
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: When creating protobufs, I prefer to have each item on a new line.

File: impl/monitor/monitor.go [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+		got.GetMapRevision() == want.GetMapRevision() {
+		// We already processed this SMR. Do not update seen SMRs. Do not scroll
+		// pages for further mutations. Return empty mutations list.
+		glog.Infof("Already processed this SMR with revision %v. Continuing.", got.GetMapRevision())
+		return nil, nil
+	}
+
+	mutations := make([]*ktpb.Mutation, pageSize*2)
+	mutations = append(mutations, resp.GetMutations()...)
+	if err := s.pageMutations(ctx, resp, mutations, opts...); err != nil {
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Is there a reason `pageMutations` isn't creating and returning the list of mutations inside of it? 
In Go, we try to avoid returning values through arguments.

File: impl/monitor/monitor.go [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+
+	// TODO(ismail): Sign reconstructed Hash instead of response hash:
+	sig, err := s.signer.SignObject(resp.GetSmr().GetRootHash())
+	if err != nil {
+		return nil, fmt.Errorf("s.signer.SignObject(_): %v", err)
+	}
+	// Update seen/processed signed map roots:
+	s.proccessedSMRs = append(s.proccessedSMRs,
+		&ktpb.GetMonitoringResponse{
+			Smr: &trillian.SignedMapRoot{
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Should this just be `resp.GetSmr()`?

File: impl/monitor/monitor.go [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+				MapRevision:    resp.GetSmr().GetMapRevision(),
+			},
+			IsValid:            true,
+			SeenTimestampNanos: seen,
+		})
+
+	return mutations, nil
+}
+
+func (s *Server) verifyMutations(ms []*ktpb.Mutation, expectedRoot []byte) error {
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: the verification logic probably deserves it's own file

File: impl/monitor/monitor.go [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+	// TODO(ismail):
+	// For each received mutation in epoch e:
+	// Verify that the provided leaf’s inclusion proof goes to epoch e -1.
+	// Verify the mutation’s validity against the previous leaf.
+	// Compute the new leaf and store the intermediate hashes locally.
+	// Compute the new root using local intermediate hashes from epoch e.
+	for _, m := range ms {
+		idx := m.GetProof().GetLeaf().GetIndex()
+		nbrs := m.GetProof().GetInclusion()
+		if err := s.tree.VerifyProof(nbrs, idx, m.GetProof().GetLeaf().GetLeafValue(),
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Try using the trillian verification logic

File: impl/monitor/monitor.go [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+		resp, err := s.client.GetMutations(ctx, req, opts...)
+		if err != nil {
+			return err
+		}
+		ms = append(ms, resp.GetMutations()...)
+	}
+	return nil
+}
+
+func (s *Server) lastSeenSMR() *trillian.SignedMapRoot {
```
Comment (gdbelvin) [inline impl/monitor/monitor.go]: Functions that implement the proto interface also probably deserve their own file

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto (hunk 14 above)
Comment (Liamsi) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: We might reuse [`SignedEntryTimestamp`](https://github.com/google/trillian/blob/938aab2b6d2c56385b3213341f264452094f4a40/trillian.proto#L175-L179) in trillian.proto

File: cmd/keytransparency-monitor/main.go [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+	mopb "github.com/google/keytransparency/impl/proto/monitor_v1_service"
+	mupb "github.com/google/keytransparency/impl/proto/mutation_v1_service"
+)
+
+var (
+	addr     = flag.String("addr", ":8099", "The ip:port combination to listen on")
+	keyFile  = flag.String("tls-key", "genfiles/server.key", "TLS private key file")
+	certFile = flag.String("tls-cert", "genfiles/server.pem", "TLS cert file")
+
+	signingKey         = flag.String("sign-key", "genfiles/p256-key.pem", "Path to private key PEM for SMH signing")
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: This binary needs separate keys!!

File: cmd/keytransparency-monitor/main.go [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+	}
+	mcc := mupb.NewMutationServiceClient(grpcc)
+
+	// Read signing key:
+	key, err := keys.NewFromPrivatePEMFile(*signingKey, *signingKeyPassword)
+	if err != nil {
+		glog.Fatalf("Could not create signer from %v: %v", *signingKey, err)
+	}
+
+	srv := monitor.New(mcc, *crypto.NewSHA256Signer(key), *mapKey, *logKey, *pollPeriod)
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: You probably want to pass a pointer to the private key so as not not have multiple copies of it lying around in memory

File: cmd/keytransparency-client/grpcc/grpc_client.go [hunk 23]
```diff
@@ -206,12 +205,7 @@ func (c *Client) Update(ctx context.Context, userID, appID string, profileData [
 		return nil, fmt.Errorf("CreateUpdateEntryRequest: %v", err)
 	}
 
-	// Check the mutation before submitting it.
-	m, err := proto.Marshal(req.GetEntryUpdate().GetUpdate())
-	if err != nil {
-		return nil, fmt.Errorf("proto.Marshal(): %v", err)
-	}
-	if _, err := c.mutator.Mutate(getResp.GetLeafProof().GetLeaf().GetLeafValue(), m); err != nil {
+	if _, err := c.mutator.Mutate(getResp.GetLeafProof().GetLeaf().GetLeafValue(), req.GetEntryUpdate().GetUpdate()); err != nil {
```
Comment (gdbelvin) [inline cmd/keytransparency-client/grpcc/grpc_client.go]: Changes to widely used APIS like `Mutate` typically touch a lot of files and are best processed as their own PRs.
Would you mind breaking this out?

File: core/mutator/entry/entry.go [hunk 24]
```diff
@@ -81,20 +76,20 @@ func (*Entry) Mutate(oldValue, mutation []byte) ([]byte, error) {
 		return nil, mutator.ErrMissingKey
 	}
 
-	if err := verifyKeys(oldValue, kv, update, entry); err != nil {
+	if err := VerifyKeys(oldValue, kv, update, entry); err != nil {
```
Comment (gdbelvin) [inline core/mutator/entry/entry.go]: I couldn't find a new external caller to `verifyKeys`.  Perhaps I'm missing something? 
Would love to keep it it a private API if it's not used externally.

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto [hunk 25]
```diff
@@ -21,10 +21,10 @@ syntax = "proto3";
 // associated with it.
 package keytransparency.v1.types;
 
-import "crypto/keyspb/keyspb.proto";
-import "crypto/sigpb/sigpb.proto";
-import "trillian.proto";
-import "trillian_map_api.proto";
+import "github.com/google/trillian/crypto/keyspb/keyspb.proto";
```
Comment (gdbelvin) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: keep the relative import scheme for consistency with the other files. 
If you're getting compile errors, `protoc` most likely needs a new `-I` import for the trillian repo

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto [hunk 26]
```diff
@@ -251,3 +251,67 @@ message GetDomainInfoResponse {
   // Vrf contains the VRF public key.
   keyspb.PublicKey vrf = 3;
 }
+
+// GetMonitoringRequest contains the input parameters of the GetMonitoring APIs.
+message GetMonitoringRequest {
```
Comment (gdbelvin) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: These guys deserve their own file since they're specific to a new service

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto (hunk 14 above)
Comment (gdbelvin) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: Brainstorming a few options:
- Update the signature field in SMR to be a set of signature timestamps. 
- Defining an equals method on `SignedMapRoot` to determine that signatures of of the same object.

File: impl/monitor/verify.go [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package monitor implements the monitor service. A monitor repeatedly polls a
+// key-transparency server's Mutations API and signs Map Roots if it could
+// reconstruct
+// clients can query.
+package monitor
```
Comment (gdbelvin) [inline impl/monitor/verify.go]: Verification functions should be in core

File: impl/proto/monitor_v1_service/monitor_v1_service.proto [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+
+import "github.com/google/keytransparency/core/proto/keytransparency_v1_types/keytransparency_v1_types.proto";
+import "google/api/annotations.proto";
+
+// The Monitor Service API allows clients to query the monitors observed and
+// validated signed map roots.
+//
+// - Signed Map Roots can be collected using the GetSignedMapRoot APIs.
+// - Monitor resources are named:
+//   /v1/map/roots/* /v1/map/roots:latest and v1/map/roots:stream.
```
Comment (gdbelvin) [inline impl/proto/monitor_v1_service/monitor_v1_service.proto]: break these resources out into separate bullet points?

File: impl/proto/mutation_v1_service/gen.go [hunk 29]
```diff
@@ -12,8 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//go:generate protoc -I=. -I=$GOPATH/src/ -I=$GOPATH/src/github.com/google/trillian/ -I=$GOPATH/src/github.com/googleapis/googleapis/ --go_out=,plugins=grpc:. mutation_v1_service.proto
```
Comment (gdbelvin) [inline impl/proto/mutation_v1_service/gen.go]: are you sure we need this change?

File: scripts/gen_monitor_keys.sh [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+  esac
+done
+
+
+DEFAULT_PWD=towel
+
+# Generate monitor signing key-pair:
+if ((INTERACTIVE == 1)); then
+  # Prompts for password:
+  openssl ecparam -name prime256v1 -genkey | openssl ec -aes256 -out p256-monitor_sign-key.pem
```
Comment (gdbelvin) [inline scripts/gen_monitor_keys.sh]: You can probably drop the p256 prefix on the key name

File: scripts/gen_server_keys.sh [hunk 31]
```diff
@@ -18,19 +18,20 @@ COMMONNAME=""
 ADDRESS=""
 SUBJECT="/C=US"
 
-while getopts d:a: option; do
+while getopts d:a:s: option; do
```
Comment (gdbelvin) [inline scripts/gen_server_keys.sh]: This change also probably deserves to be it's own PR too

File: core/mutator/entry/entry.go (hunk 24 above)
Comment (Liamsi) [inline core/mutator/entry/entry.go]: Thanks. I started with exporting and calling the checks which happen in `Mutate`. It makes more sense to just call `Mutate` (which does the checks and return the new leaf which the monitor needs anyways).

File: scripts/gen_server_keys.sh (hunk 31 above)
Comment (Liamsi) [inline scripts/gen_server_keys.sh]: There already is a (merged) PR for that: https://github.com/google/keytransparency/pull/725
Will clean it in this branch. Sorry for the noise.

File: cmd/keytransparency-client/grpcc/grpc_client.go (hunk 23 above)
Comment (Liamsi) [inline cmd/keytransparency-client/grpcc/grpc_client.go]: Thanks, sure here is a separate PR for this: https://github.com/google/keytransparency/pull/739

File: cmd/keytransparency-monitor/main.go [hunk 32]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+func dial(ktURL, caFile string) (*grpc.ClientConn, error) {
+	var opts []grpc.DialOption
+	host, _, err := net.SplitHostPort(ktURL)
+	if err != nil {
+		return nil, err
+	}
+	var creds credentials.TransportCredentials
+	if caFile != "" {
```
Comment (gdbelvin) [inline cmd/keytransparency-monitor/main.go]: follow the example in the key transparency client, splitting up `getTransportCreds`

File: core/monitor/monitor.go [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+
+type Monitor struct {
+	hasher      hashers.MapHasher
+	logPubKey   crypto.PublicKey
+	mapPubKey   crypto.PublicKey
+	logVerifier merkle.LogVerifier
+	trusted     trillian.SignedLogRoot
+}
+
+func New(hasher hashers.MapHasher, mapPubKey, logPubKey crypto.PublicKey, logVerifier merkle.LogVerifier) *Monitor {
```
Comment (gdbelvin) [inline core/monitor/monitor.go]: Maybe pass in the tree structs and compute the rest locally

File: core/monitor/verify.go [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	return resp
+}
+
+func verifyMutations(muts []*ktpb.Mutation, expectedRoot []byte, mapID int64) error {
+	newLeaves := make([]merkle.HStar2LeafHash, 0, len(muts))
+	mutator := entry.New()
+	oldProofNodes := make(map[string][]byte)
+	hasher := coniks.Default
```
Comment (gdbelvin) [inline core/monitor/verify.go]: Pull from tree config

File: core/monitor/verify.go [hunk 35]
```diff
@@ -0,0 +1,171 @@
...
+		newLeafnID := storage.NewNodeIDFromPrefixSuffix(index, storage.Suffix{}, hasher.BitLen())
+		newLeafHash := hasher.HashLeaf(mapID, index, newLeaf)
+		newLeaves = append(newLeaves, merkle.HStar2LeafHash{
+			Index:    newLeafnID.BigInt(),
+			LeafHash: newLeafHash,
+		})
+		// store the proof hashes locally to recompute the tree below:
+		sibIDs := newLeafnID.Siblings()
+		for level, proof := range m.GetProof().GetInclusion() {
+			pID := sibIDs[level]
```
Comment (gdbelvin) [inline core/monitor/verify.go]: Level is not a good enough index.

Comment (gdbelvin) [inline core/monitor/verify.go]: try iterating over sibIDs

File: scripts/gen_monitor_keys.sh (hunk 30 above)
Comment (Liamsi) [inline scripts/gen_monitor_keys.sh]: done

File: impl/proto/monitor_v1_service/monitor_v1_service.proto (hunk 28 above)
Comment (Liamsi) [inline impl/proto/monitor_v1_service/monitor_v1_service.proto]: done

File: impl/monitor/verify.go (hunk 27 above)
Comment (Liamsi) [inline impl/monitor/verify.go]: done

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto (hunk 26 above)
Comment (Liamsi) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: done

File: core/proto/keytransparency_v1_types/keytransparency_v1_types.proto (hunk 25 above)
Comment (Liamsi) [inline core/proto/keytransparency_v1_types/keytransparency_v1_types.proto]: done

File: core/mutator/entry/entry.go (hunk 24 above)
Comment (Liamsi) [inline core/mutator/entry/entry.go]: Change back to private.

File: cmd/keytransparency-monitor/main.go (hunk 22 above)
Comment (Liamsi) [inline cmd/keytransparency-monitor/main.go]: done

File: cmd/keytransparency-monitor/main.go (hunk 21 above)
Comment (Liamsi) [inline cmd/keytransparency-monitor/main.go]: done

File: impl/monitor/monitor.go (hunk 19 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: Will be moved out of this PR. But done

File: impl/monitor/monitor.go (hunk 18 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: done

File: impl/proto/mutation_v1_service/gen.go (hunk 29 above)
Comment (Liamsi) [inline impl/proto/mutation_v1_service/gen.go]: Hmm, this seems to be neccessary on my machine to make sure that go generate uses the right version of googleapis. 
If I remove this line all calls to `grpc.Errorf` in the generated code get replaced with `status.Errorf` for instance. And I get error ala:
```
impl/proto/mutation_v1_service/mutation_v1_service.pb.gw.go:165: too many arguments in call to runtime.AnnotateContext
        have ("golang.org/x/net/context".Context, *runtime.ServeMux, *http.Request)
        want ("golang.org/x/net/context".Context, *http.Request)
``` 
while compiling. /cc @cesarghali

File: impl/monitor/monitor.go (hunk 6 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: removed the stream API as I'm definitely not finishing with this PR.

File: impl/monitor/monitor.go (hunk 16 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: done

File: impl/monitor/monitor.go (hunk 15 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: done

File: impl/monitor/monitor.go (hunk 17 above)
Comment (Liamsi) [inline impl/monitor/monitor.go]: As discussed there will be 2 PRs. One monitor that doesn't do anything and just signs what it sees and a separate pull for the verification logic.

File: impl/proto/mutation_v1_service/gen.go (hunk 29 above)
Comment (Liamsi) [inline impl/proto/mutation_v1_service/gen.go]: I'll simply remove this change as I don't need to change the mutation service proto?

**Comment by Liamsi** [comment]
Closing this PR in favor of #776 and #768
